{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NC_035107.1': [51804503, 103609007, 155413511, 207218014, 259022518], 'NC_035108.1': [79070952, 158141905, 237212858, 316283810, 395354763], 'NC_035109.1': [68296278, 136592556, 204888835, 273185113, 341481391], 'NC_035159.1': [2798, 5596, 8395, 11193, 13991]}\n",
      "NC_035107.1\n",
      "4\n",
      "NC_035107.1\n",
      "3\n",
      "NC_035107.1\n",
      "2\n",
      "NC_035107.1\n",
      "1\n",
      "NC_035107.1\n",
      "0\n",
      "NC_035108.1\n",
      "4\n",
      "NC_035108.1\n",
      "3\n",
      "NC_035108.1\n",
      "2\n",
      "NC_035108.1\n",
      "1\n",
      "NC_035108.1\n",
      "0\n",
      "NC_035109.1\n",
      "4\n",
      "NC_035109.1\n",
      "3\n",
      "NC_035109.1\n",
      "2\n",
      "NC_035109.1\n",
      "1\n",
      "NC_035109.1\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "from os.path import expanduser\n",
    "\n",
    "##IF blastDB does not yet exist, make one using:\n",
    "wd=expanduser(\"~/Mosquitoes/ResistanceGenes/\") #Change the directory as appropriate\n",
    "gffFile=wd+\"/InputData/GCF_002204515.2_AaegL5.0_genomic.gff\" #Get these from NCBI/ENA\n",
    "fastaFile=wd+\"/InputData/GCF_002204515.2_AaegL5.0_genomic.fna\"  #Get these from NCBI/ENA\n",
    "\n",
    "\n",
    "blastCommand=f'makeblastdb -in {fastaFile} -title AaegL5 -out {wd}/blastDB/AaegL5 -dbtype nucl  -blastdb_version 4 '\n",
    "subprocess.run(blastCommand, shell=True, executable=\"/bin/bash\", stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "\n",
    "\n",
    "\n",
    "chrLength={}\n",
    "exons={}\n",
    "with open(gffFile) as file:\n",
    "    for line in file:\n",
    "        if line[0:2]==\"NC\":\n",
    "            #these are the chromosomes, unplaced contigs start with NW\n",
    "            values=line.split(\"\\t\")\n",
    "            if values[2]==\"region\":\n",
    "               chrLength[values[0]]=int(values[4]) \n",
    "\n",
    "targetmRNA={}\n",
    "targetGenePositions={}\n",
    "for chr in chrLength:\n",
    "    targetGenePositions[chr]=[]\n",
    "    targetmRNA[chr]=[]\n",
    "    for i in range(1,6):\n",
    "        targetGenePositions[chr].append( int(chrLength[chr] / 6 * i) )\n",
    "\n",
    "\n",
    "def GetBlastMatchesLength(fastaSequence):\n",
    "    with open(f'{wd}/temp.fasta', \"w\") as outputFasta:  \n",
    "        outputFasta.write(\">ExonSeqeuence\\n\"+fastaSequence+\"\\n\")\n",
    "    blastCommand=\"blastn -query \"+wd+\"/temp.fasta -task 'megablast'  -max_target_seqs 1000 -db \"+wd+\"/blastDB/AaegL5 -num_threads 8 -evalue 0.01 -word_size 28 -outfmt '6 delim=  qseqid qstart qend sseqid sstart send pident evalue'\"\n",
    "    gff=subprocess.run(blastCommand, shell=True, executable=\"/bin/bash\", stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    result=0\n",
    "    for line in gff.stdout.decode().split(\"\\n\"):\n",
    "        if len(line)>0:\n",
    "            result+=int(line.split(\"\\t\")[2])-int(line.split(\"\\t\")[1])\n",
    "    return result\n",
    "\n",
    "def GetAnnotationValue(targetAnnotationLabel, wholeAnnotationField):\n",
    "    annotations=wholeAnnotationField.strip().split(\";\")\n",
    "    for annotation in annotations:\n",
    "        annLabel,annValue=annotation.split(\"=\")\n",
    "        if annLabel==targetAnnotationLabel:\n",
    "            return annValue\n",
    "    return \"AnnotationNotFound\"\n",
    "\n",
    "def CollectExonSequences(chromosome, mRNAid):\n",
    "    targetExons=[] # [(exon1 start, exon1 end), (exon2 start, exon2 end)]\n",
    "    with open(gffFile) as file:\n",
    "        for line in file:\n",
    "            if line[0]!=\"#\":\n",
    "                values=line.split(\"\\t\")\n",
    "                if values[0]==chromosome and values[2]==\"exon\":\n",
    "                    transcriptID=GetAnnotationValue(\"transcript_id\", values[8].strip())\n",
    "                    if mRNAid==transcriptID:\n",
    "                        targetExons.append( (int(values[3]), int(values[4])) )\n",
    "    exonSequence=\"\"\n",
    "    for chr in SeqIO.parse(fastaFile, \"fasta\"):\n",
    "        if chr.id in chromosome:\n",
    "            for exonStart, exonEnd in targetExons:\n",
    "                exonSequence=exonSequence+str(chr.seq[exonStart:exonEnd])\n",
    "    return exonSequence\n",
    "\n",
    "print(targetGenePositions)\n",
    "with open(wd+\"resistanceGeneExonSequences.fasta\",\"w\") as exonOutputFile:\n",
    "    with open(gffFile) as file:\n",
    "        for line in file:\n",
    "            if line[0]!=\"#\":\n",
    "                values=line.split(\"\\t\")\n",
    "                if values[2]==\"mRNA\" and values[0] in targetGenePositions and len(targetGenePositions[values[0]])>0:\n",
    "                    if int(values[4])-int(values[3])>5000 and int(values[4])-int(values[3])<25000:\n",
    "                        #check if gene coordinates are just after the first target chromosomes position\n",
    "                        #add mRNA and remove the target position\n",
    "                        if int(values[3])>targetGenePositions[values[0]][0]:\n",
    "                            #get transcriptid:\n",
    "                            transcriptID=GetAnnotationValue(\"transcript_id\", values[8].strip())\n",
    "                            trascriptExonsSequence=CollectExonSequences(values[0],transcriptID)\n",
    "                            blastTotalLength=GetBlastMatchesLength(trascriptExonsSequence)\n",
    "                            if abs(blastTotalLength-len(trascriptExonsSequence)) < 10:\n",
    "                                targetmRNA[values[0]].append(line)\n",
    "                                targetGenePositions[values[0]].pop(0)\n",
    "                                exonOutputFile.write(\">\"+transcriptID+\"\\n\"+trascriptExonsSequence+\"\\n\")\n",
    "                                print(values[0])\n",
    "                                print(len(targetGenePositions[values[0]]))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='bedtools getfasta -name  -fi ~/Mosquitoes/GCF_002204515.2_AaegL5.0_genomic.fna -bed ~/Mosquitoes/temp.bed > ~/Mosquitoes/ResistanceGenes/genesV2.fasta ', returncode=0)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promoterBuffer=300\n",
    "\n",
    "resitanceGenes=[(\"XM_021851332.1\",161486025,161645623,\"ACE\",\"NC_035109.1\"),\n",
    "                (\"XM_021840622.1\", 41628786,41861946,\"GABA\",\"NC_035108.1\"),\n",
    "                (\"XM_021852340.1\",315926360,316405638,\"VSSC\",\"NC_035109.1\"),\n",
    "                (\"XM_021846286.1\",351633367,351634957,\"GSTE2\",\"NC_035108.1\"),\n",
    "                (\"YP_009389261.1\",1298,2834,\"COX1\",\"NC_035159.1\")]\n",
    "\n",
    "\n",
    "def write_to_bed(line, file_handle):\n",
    "        values=line.split(\"\\t\")\n",
    "        values[3]=str(int(values[3])-promoterBuffer)\n",
    "        values[4]=str(int(values[4])-promoterBuffer)\n",
    "        file_handle.write( \"\\t\".join( [values[0], values[3], values[4],GetAnnotationValue(\"transcript_id\", values[8].strip())] ) +\"\\n\")\n",
    "\n",
    "with open(f'{wd}/InputData/mappedRegions.bed', \"w\") as bed_output:        \n",
    "    for chr, target in targetmRNA.items():\n",
    "        for item in targetmRNA[chr]:\n",
    "             write_to_bed(item, bed_output)\n",
    "    for value in resitanceGenes:\n",
    "         bed_output.write(\"\\t\".join(  [value[4], str(value[1]-promoterBuffer), str(value[2]+promoterBuffer), value[0] ]  )+\"\\n\")\n",
    "\n",
    "subprocess.run(\"bedtools getfasta -name  -fi \"+fastaFile+\" -bed \"+wd+\"/InputData/mappedRegions.bed > \"+wd+\"/InputData/mappedRegions.fasta \", shell=True, executable=\"/bin/bash\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "a612c00234cb2afc6511c78aaf3593a4092443bb2a0bed4620e32910b743a5d1"
  },
  "kernelspec": {
   "display_name": "Python 3.7.6 ('base': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
